src/object.d

```diff
@@ -2309,6 +2310,94 @@ inout(V) get(K, V)(inout(V[K])* aa, K key, lazy inout(V) defaultValue)
     }
 }
 
+unittest
+{
+    static class T
+    {
+        static size_t count;
+        this() { ++count; }
+    }
+
+    T[string] aa;
+
+    auto a = new T;
+    aa["foo"] = a;
+    assert(T.count == 1);
+    auto b = aa.get("foo", new T);
+    assert(T.count == 1);
+    assert(b is a);
+    auto c = aa.get("bar", new T);
+    assert(T.count == 2);
+    assert(c !is a);
+
+    //Obviously get doesn't add.
+    assert("bar" !in aa);
+}
+
+ref V getOrAdd(K, V)(ref V[K] aa, K key, lazy V value = V.init) pure
+{
+    bool found;
+    auto p = cast(V*) _aaGetX(cast(void**)&aa, typeid(V[K]), &key, found);
+    return found ? *p : (*p = value);
+}
+
+unittest
+{
+    static class T
+    {
+        static size_t count;
+        this() { ++count; }
+    }
+
+    T[string] aa;
+
+    auto a = new T;
+    aa["foo"] = a;
+    assert(T.count == 1);
+    auto b = aa.getOrAdd("foo", new T);
+    assert(T.count == 1);
+    assert(b is a);
+    auto c = aa.getOrAdd("bar", null);
+    assert(T.count == 1);
+    assert(c is null);
+    auto d = aa.getOrAdd("bar", new T);
+    assert(d is null);
+    auto e = aa.getOrAdd("baz", new T);
+    assert(T.count == 2);
+    assert(e !is a);
+
+    assert("baz" in aa);
+}
+
+unittest
+{
+    struct S
```

Comment:
static struct here, otherwise there is a connection between the unittest stack frame and every S instance.